### ISSUE TITLE
pub upgrade throws "Bad state: Cannot add new events after calling close"

### ISSUE THREAD

**Issue by zoechi** [dart-lang/pub#1407]
running `pub upgrade` produces

```
(master) $ pub upgrade --trace
Resolving dependencies... (19.5s) 
  _discoveryapis_commons 0.1.3+1
  analyzer 0.27.3
  ansicolor 0.0.9
! appengine 0.3.3 (overridden)
  args 0.13.4
  async 1.10.0
  barback 0.15.2+7
  bignum 0.0.7 (0.1.0 available)
  boolean_selector 1.0.1
  bwu_grinder_tasks 0.1.1-dev.3
  bwu_log 0.1.0
  charcode 1.1.0
  cipher 0.7.1
  cli_util 0.0.1+2
  collection 1.5.1
  concepts 0.2.0
  convert 1.0.1
! crypto 1.1.0 (overridden)
  csslib 0.12.2
  dart_jwt 0.4.6
  either 0.1.8
  fixnum 0.9.1+2 (0.10.2 available)
! gcloud 0.2.0+12 (overridden)
  glob 1.1.2
! googleapis 0.25.0 (overridden)
! googleapis_auth 0.2.3+2 (overridden)
! googleapis_beta 0.25.0 (overridden)
  grinder 0.8.0+2
  html 0.12.2+1
  http 0.11.3+3
  http_exception 0.1.0
  http_multi_server 2.0.1
  http_parser 2.2.1
  http_server 0.9.5+1
  intl 0.12.7+1
  logging 0.11.2
  logging_handlers 0.8.0
  matcher 0.12.0+2
  memcache 0.2.0
  mime 0.9.3
  mockable_filesystem 0.0.3
  option 1.1.0
  package_config 0.1.3
  path 1.3.9
  petitparser 1.5.3
  plugin 0.1.0
  pool 1.2.3
  protobuf 0.5.1+1
  pub_semver 1.2.4
  quiver 0.21.4
  shelf 0.6.5
  shelf_appengine 0.2.3
  shelf_auth 0.7.2
  shelf_path 0.1.8
  shelf_route 0.14.2
  shelf_static 0.2.3+3
  shelf_web_socket 0.2.0
  source_map_stack_trace 1.0.4
  source_maps 0.10.1+1
  source_span 1.2.2
  stack_trace 1.6.5
  stream_channel 1.3.1
  string_scanner 0.1.4+1
  supports_color 0.1.1
  test 0.12.13
  typed_data 1.1.2
  unscripted 0.6.2
  uri 0.11.0
  utf 0.9.0+3
  uuid 0.5.2
  watcher 0.9.7
  web_socket_channel 1.0.2
  when 0.2.0
  which 0.1.3
  yaml 2.1.8
Warning: You are using these overridden dependencies:
! appengine 0.3.3
! crypto 1.1.0
! gcloud 0.2.0+12
! googleapis 0.25.0
! googleapis_auth 0.2.3+2
! googleapis_beta 0.25.0
No dependencies changed.
Precompiling executables... 
Loading source assets... 
ERROR - 2016-04-17 13:09:18.752558
GET /generate_from_arb.dart
Error thrown by handler.
Bad state: Cannot add new events after calling close
dart:async                                             _BroadcastStreamController.addError
package:pub/src/barback/base_server.dart 117           BaseServer._handleError
package:shelf/src/middleware.dart 60                   createMiddleware.<fn>
dart:async                                             _Completer.completeError
package:pub/src/barback/barback_server.dart 207        BarbackServer._serveAsset.<async>
===== asynchronous gap ===========================
package:shelf/src/middleware.dart 69                   createMiddleware.<fn>.<fn>.<fn>
===== asynchronous gap ===========================
package:shelf/shelf_io.dart 52                         serveRequests.<fn>.<fn>
===== asynchronous gap ===========================
package:shelf/shelf_io.dart 51                         serveRequests
package:pub/src/barback/base_server.dart 43            BaseServer.BaseServer
package:pub/src/barback/barback_server.dart 69         BarbackServer.BarbackServer._
package:pub/src/barback/barback_server.dart 63         BarbackServer.bind.<fn>
dart:async                                             _SyncCompleter.complete
package:pub/src/io.dart 868                            bindServer.<async>
===== asynchronous gap ===========================
dart:async                                             _Future.then
package:pub/src/barback/barback_server.dart 57         BarbackServer.bind
package:pub/src/barback/asset_environment.dart 251     AssetEnvironment.servePackageBinDirectory.<fn>
===== asynchronous gap ===========================
dart:async                                             _Future.then
package:pub/src/barback/asset_environment.dart 250     AssetEnvironment.servePackageBinDirectory
package:pub/src/barback/asset_environment.dart 271     AssetEnvironment.precompileExecutables.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/barback/asset_environment.dart         AssetEnvironment.precompileExecutables
package:pub/src/entrypoint.dart 356                    Entrypoint.precompileExecutables.<async>.<fn>.<async>.<fn>.<async>
===== asynchronous gap ===========================
dart:async                                             Future.wait
package:pub/src/utils.dart 149                         waitAndPrintErrors
package:pub/src/entrypoint.dart 353                    Entrypoint.precompileExecutables.<async>.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 83      AssetEnvironment.create.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 516     AssetEnvironment._load.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 671     AssetEnvironment._withStreamErrors.<fn>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 514     AssetEnvironment._load.<fn>.<async>.<fn>.<async>.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/load_all_transformers.dart 44  loadAllTransformers.<async>
===== asynchronous gap ===========================
package:pub                                            Entrypoint.precompileExecutables.<async>.<fn>
package:pub/src/log.dart 407                           progress
package:pub/src/entrypoint.dart 333                    Entrypoint.precompileExecutables.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/entrypoint.dart                        Entrypoint.precompileExecutables
package:pub/src/entrypoint.dart 205                    Entrypoint.acquireDependencies.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/entrypoint.dart 231                    Entrypoint._precompileDependencies.<async>
===== asynchronous gap ===========================
dart:async                                             _asyncThenWrapperHelper
package:pub/src/entrypoint.dart                        Entrypoint.acquireDependencies
package:pub/src/command/upgrade.dart 34                UpgradeCommand.run.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/command/upgrade.dart                   UpgradeCommand.run
package:args/command_runner.dart 176                   CommandRunner.runCommand.<fn>
dart:async                                             Future.Future.sync
package:args/command_runner.dart 129                   CommandRunner.runCommand
package:pub/src/command_runner.dart 131                PubCommandRunner.runCommand.<async>.<fn>
dart:async                                             Future.Future.sync
package:pub/src/utils.dart 113                         captureErrors.<fn>
package:stack_trace                                    Chain.capture
package:pub/src/utils.dart 127                         captureErrors
package:pub/src/command_runner.dart 131                PubCommandRunner.runCommand.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/command_runner.dart 217                PubCommandRunner._validatePlatform.<async>

ERROR - 2016-04-17 13:09:18.766877
GET /rewrite_intl_messages.dart
Error thrown by handler.
Bad state: Cannot add new events after calling close
dart:async                                             _BroadcastStreamController.addError
package:pub/src/barback/base_server.dart 117           BaseServer._handleError
package:shelf/src/middleware.dart 60                   createMiddleware.<fn>
dart:async                                             _Completer.completeError
package:pub/src/barback/barback_server.dart 207        BarbackServer._serveAsset.<async>
===== asynchronous gap ===========================
package:shelf/src/middleware.dart 69                   createMiddleware.<fn>.<fn>.<fn>
===== asynchronous gap ===========================
package:shelf/shelf_io.dart 52                         serveRequests.<fn>.<fn>
===== asynchronous gap ===========================
package:shelf/shelf_io.dart 51                         serveRequests
package:pub/src/barback/base_server.dart 43            BaseServer.BaseServer
package:pub/src/barback/barback_server.dart 69         BarbackServer.BarbackServer._
package:pub/src/barback/barback_server.dart 63         BarbackServer.bind.<fn>
dart:async                                             _SyncCompleter.complete
package:pub/src/io.dart 868                            bindServer.<async>
===== asynchronous gap ===========================
dart:async                                             _Future.then
package:pub/src/barback/barback_server.dart 57         BarbackServer.bind
package:pub/src/barback/asset_environment.dart 251     AssetEnvironment.servePackageBinDirectory.<fn>
===== asynchronous gap ===========================
dart:async                                             _Future.then
package:pub/src/barback/asset_environment.dart 250     AssetEnvironment.servePackageBinDirectory
package:pub/src/barback/asset_environment.dart 271     AssetEnvironment.precompileExecutables.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/barback/asset_environment.dart         AssetEnvironment.precompileExecutables
package:pub/src/entrypoint.dart 356                    Entrypoint.precompileExecutables.<async>.<fn>.<async>.<fn>.<async>
===== asynchronous gap ===========================
dart:async                                             Future.wait
package:pub/src/utils.dart 149                         waitAndPrintErrors
package:pub/src/entrypoint.dart 353                    Entrypoint.precompileExecutables.<async>.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 83      AssetEnvironment.create.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 516     AssetEnvironment._load.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 671     AssetEnvironment._withStreamErrors.<fn>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/asset_environment.dart 514     AssetEnvironment._load.<fn>.<async>.<fn>.<async>.<fn>.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/barback/load_all_transformers.dart 44  loadAllTransformers.<async>
===== asynchronous gap ===========================
package:pub                                            Entrypoint.precompileExecutables.<async>.<fn>
package:pub/src/log.dart 407                           progress
package:pub/src/entrypoint.dart 333                    Entrypoint.precompileExecutables.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/entrypoint.dart                        Entrypoint.precompileExecutables
package:pub/src/entrypoint.dart 205                    Entrypoint.acquireDependencies.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/entrypoint.dart 231                    Entrypoint._precompileDependencies.<async>
===== asynchronous gap ===========================
dart:async                                             _asyncThenWrapperHelper
package:pub/src/entrypoint.dart                        Entrypoint.acquireDependencies
package:pub/src/command/upgrade.dart 34                UpgradeCommand.run.<async>
===== asynchronous gap ===========================
dart:async                                             Future.Future.microtask
package:pub/src/command/upgrade.dart                   UpgradeCommand.run
package:args/command_runner.dart 176                   CommandRunner.runCommand.<fn>
dart:async                                             Future.Future.sync
package:args/command_runner.dart 129                   CommandRunner.runCommand
package:pub/src/command_runner.dart 131                PubCommandRunner.runCommand.<async>.<fn>
dart:async                                             Future.Future.sync
package:pub/src/utils.dart 113                         captureErrors.<fn>
package:stack_trace                                    Chain.capture
package:pub/src/utils.dart 127                         captureErrors
package:pub/src/command_runner.dart 131                PubCommandRunner.runCommand.<async>
dart:async                                             _SyncCompleter.complete
package:pub/src/command_runner.dart 217                PubCommandRunner._validatePlatform.<async>

Failed to precompile shelf_appengine:server:
Unhandled exception:
Could not load "http://localhost:42120/server.dart": Failure getting http://localhost:42120/server.dart:
  500 Internal Server Error
#0      _asyncLoadErrorCallback (dart:_builtin:162)
#1      _asyncLoadError (dart:_builtin:584)
#2      _handleLoaderReply (dart:_builtin:401)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
Failed to precompile intl:extract_to_arb:
Unhandled exception:
Could not load "http://localhost:38703/extract_to_arb.dart": Failure getting http://localhost:38703/extract_to_arb.dart:
  500 Internal Server Error
#0      _asyncLoadErrorCallback (dart:_builtin:162)
#1      _asyncLoadError (dart:_builtin:584)
#2      _handleLoaderReply (dart:_builtin:401)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
Failed to precompile intl:generate_from_arb:
Unhandled exception:
Could not load "http://localhost:38703/generate_from_arb.dart": Failure getting http://localhost:38703/generate_from_arb.dart:
  500 Internal Server Error
#0      _asyncLoadErrorCallback (dart:_builtin:162)
#1      _asyncLoadError (dart:_builtin:584)
#2      _handleLoaderReply (dart:_builtin:401)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
Failed to precompile intl:rewrite_intl_messages:
Unhandled exception:
Could not load "http://localhost:38703/rewrite_intl_messages.dart": Failure getting http://localhost:38703/rewrite_intl_messages.dart:
  500 Internal Server Error
#0      _asyncLoadErrorCallback (dart:_builtin:162)
#1      _asyncLoadError (dart:_builtin:584)
#2      _handleLoaderReply (dart:_builtin:401)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
Failed to precompile test:test:
Unhandled exception:
Could not load "http://localhost:34633/test.dart": Failure getting http://localhost:34633/test.dart:
  500 Internal Server Error
#0      _asyncLoadErrorCallback (dart:_builtin:162)
#1      _asyncLoadError (dart:_builtin:584)
#2      _handleLoaderReply (dart:_builtin:401)
#3      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:148)
```

**Comment by zoechi** [comment]
Moving back to 

> Dart VM version: 1.16.0-edge.dacf846df0447de2d4809a4f619f3ef00cbc1830 (Tue Apr 12 16:31:03 2016) on "linux_x64"

fixes it. Probably a VM issue then.

**Comment by mit-mit** [comment]
@nex3 @iposva-google , any idea what is going on here?

**Comment by zoechi** [comment]
I can't reproduce with

> Dart VM version: 1.16.0-edge.dacf846df0447de2d4809a4f619f3ef00cbc1830 (Tue Apr 12 16:31:03 2016) on "linux_x64"

With the newest bleeding edge I get dart-lang/sdk#26295

**Comment by nex3** [comment]
It looks like this can be closed thanks to 04ff0cc2cb6a3b698159b1aeef14cd3d7c90e287.
